src/auth/executable-response.ts

```diff
@@ -0,0 +1,278 @@
+        "Executable response must contain a 'success' field."
+      );
+    }
+
+    this.version = responseJson.version;
+    this.success = responseJson.success;
+
+    // Validate required fields for a successful response.
+    if (this.success) {
+      this.expirationTime = responseJson.expiration_time;
+      this.tokenType = responseJson.token_type;
+
+      // Validate token type field.
+      if (
+        this.tokenType !== SAML_SUBJECT_TOKEN_TYPE &&
+        this.tokenType !== OIDC_SUBJECT_TOKEN_TYPE1 &&
+        this.tokenType !== OIDC_SUBJECT_TOKEN_TYPE2
+      ) {
+        throw new InvalidTokenTypeFieldError(
+          "Executable response must contain a 'token_type' field when successful " +
+            `and it must be one of ${OIDC_SUBJECT_TOKEN_TYPE1}, ${OIDC_SUBJECT_TOKEN_TYPE2}, or ${SAML_SUBJECT_TOKEN_TYPE}.`
+        );
+      }
+
+      // Validate subject token.
+      if (this.tokenType === SAML_SUBJECT_TOKEN_TYPE) {
+        if (!responseJson.saml_response) {
+          throw new InvalidSubjectTokenError(
+            `Executable response must contain a 'saml_response' field when token_type=${SAML_SUBJECT_TOKEN_TYPE}.`
+          );
+        }
+        this.subjectToken = responseJson.saml_response;
+      } else {
+        if (!responseJson.id_token) {
+          throw new InvalidSubjectTokenError(
+            "Executable response must contain a 'id_token' field when " +
+              `token_type=${OIDC_SUBJECT_TOKEN_TYPE1} or ${OIDC_SUBJECT_TOKEN_TYPE2}.`
+          );
+        }
+        this.subjectToken = responseJson.id_token;
+      }
+    } else {
+      // Both code and message must be provided for unsuccessful responses.
+      if (!responseJson.code) {
+        throw new InvalidCodeFieldError(
+          "Executable response must contain a 'code' field when unsuccessful."
+        );
+      }
+      if (!responseJson.message) {
+        throw new InvalidMessageFieldError(
+          "Executable response must contain a 'message' field when unsuccessful."
+        );
+      }
+      this.errorCode = responseJson.code;
+      this.errorMessage = responseJson.message;
+    }
+  }
+
+  /**
+   * @return A boolean representing if the response has a valid token. Returns
+   * true when the response was successful and the token is not expired.
+   */
+  isValid(): boolean {
+    return !this.isExpired() && this.success;
+  }
+
+  /**
+   * @return A boolean representing if the response is expired. Returns true if the
+   * provided timeout has passed.
+   */
+  isExpired(): boolean {
+    return (
+      this.expirationTime !== undefined &&
+      this.expirationTime < Math.round(Date.now() / 1000)
+    );
+  }
+}
+
+/**
+ * An error thrown by the ExecutableResponse class.
+ */
+export class ExecutableResponseError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = 'ExecutableResponseError';
+    Object.setPrototypeOf(this, ExecutableResponseError.prototype);
+  }
```

Comment:
I like the use of descriptive error classes, however this constructor, and similar constructors (e.g. `InvalidVersionFieldError`, `InvalidSuccessFieldError`) shouldn't be necessary:
- `name` isn't referenced outside of the class
- `setPrototypeOf` doesn't supply additional functionality
- if the above 2 are removed, the default constructor contains the `super` call